### PR TITLE
`'$this'` as `$sourcePropertyName` means complete source object

### LINKED ISSUE #58: Allow converting object itself by ConvertingPopulator
We have the situation that not a certain property of the `source` object should be populated into a `target` property but the `source` object itself.

This could be easily done by skipping `$sourcePropertyName` attribute value and use the `source` object.

To implement that this issue has been created.

### PR DESCRIPTION
**Pull request by mike4git**
As mentioned in the issue:
Sometimes you want to populate the complete (source) object in a certain target attribute/property.
With that PR it is quite simple possible.

Closes #58

### CODE REVIEW

**Review by nsd0sneumann — APPROVED** [review]

**Review by nsd0sneumann — APPROVED** [review]

**Review by nsd0sneumann — APPROVED** [review]

**Review by nsd0sneumann — APPROVED** [review]

File: docs/usage.md [hunk 1]
```diff
@@ -272,6 +272,10 @@ person.address.populator:
 Be aware - that both properties have the same name should not lead you think they have the same type.
 There is really an object conversion behind done by `address.converter`.
 
+If you set the `sourcePropertyName` as empty string the complete `source` object will be used for population.
```
Comment (jdreesen) [inline docs/usage.md]:
```suggestion
If you specify the `sourcePropertyName` as an empty string, the full `source` object is used for the population.
```

File: docs/usage.md [hunk 2]
```diff
@@ -272,6 +272,10 @@ person.address.populator:
...
+
+Especially in addition wih the `ConvertingPopulator` this is sometimes necessary.
```
Comment (jdreesen) [inline docs/usage.md]:
```suggestion
Especially in connection with the `ConvertingPopulator` this is sometimes necessary.
```

But... do we really need this sentence? This whole paragraph is about the `ConvertingPopulator`.

Maybe we should rather say instead that this is only possible with the `PropertyMappingPopulator`.

File: src/Populator/PropertyMappingPopulator.php [hunk 3]
```diff
@@ -41,10 +41,16 @@ public function __construct(
     public function populate(object $target, object $source, object $ctx = null): void
     {
         try {
+            if (!empty($this->sourceProperty)) {
```
Comment (jdreesen) [inline src/Populator/PropertyMappingPopulator.php]: `empty()` should be avoided.

```suggestion
            if ('' !== $this->sourceProperty) {
```

File: tests/Populator/PropertyMappingPopulatorTest.php [hunk 4]
```diff
@@ -24,4 +25,20 @@ public function test_populate(): void
...
+            ->setStreet('Street')
+            ->setStreetNo('1')
+            ->setCity('Capitol City')
+            ->setPostalCode('12345');
+
+        $person = new Person();
+
+        $populator->populate($person, $address);
+
+        self::assertEquals('Street', $person->getAddress()->getStreet());
```
Comment (jdreesen) [inline tests/Populator/PropertyMappingPopulatorTest.php]:
```suggestion
        self::assertSame($address, $person->getAddress());
```

**Comment by jdreesen** [comment]
This PR would also mean, that with the following config:

```yaml
# config/packages/neusta_converter.yaml
neusta_converter:
  converter:
    person.converter:
      # ...
      properties:
        email: ~
        phone: ''
```

the email would be populated from the target field `email` to the source field `email` 
while the for phone, the source object would be populated into the target field `phone`. 

Would that be correct, and is this what we want?

I'd rather suggest that both `~` and `''` should do the same here.
What do you think?

File: src/Populator/PropertyMappingPopulator.php (hunk 3 above)
Comment (mike4git) [reply src/Populator/PropertyMappingPopulator.php]: This was new for me but I've seen that:
```
empty("0") is true
```
and that's of coure a no-go.

**Comment by mike4git** [comment]
> This PR would also mean, that with the following config:
> 
> ```yaml
> # config/packages/neusta_converter.yaml
> neusta_converter:
>   converter:
>     person.converter:
>       # ...
>       properties:
>         email: ~
>         phone: ''
> ```
> 
> the email would be populated from the target field `email` to the source field `email` while the for phone, the source object would be populated into the target field `phone`.
> 
> Would that be correct, and is this what we want?
> 
> I'd rather suggest that both `~` and `''` should do the same here. What do you think?

I see your point but for some circumstances it is necessary that the "complete" source object should be populated into a certain target property.
Should we write a new config for that?

What about a certain Populator type?!

**Comment by jdreesen** [comment]
> I see your point but for some circumstances it is necessary that the "complete" source object should be populated into a certain target property.

But do you need it at this place? If I understood correctly, you wanted to use this especially together with the `ConvertingPopulator`?

**Comment by mike4git** [comment]
What would you think about the following syntax:
```
# config/packages/neusta_converter.yaml
neusta_converter:
  converter:
    person.converter:
      # ...
      properties:
        email: ~
        phone: this
```
So, as you mentioned:

The complete source object will be converted into phone target property.

**Comment by jdreesen** [comment]
I really like this idea! 

But I would suggest using `$this` as the keyword. Then it is easier to recognize that it is not a normal field name.

**Review by nsd0sneumann — APPROVED** [review]

**Review by nsd0sneumann — APPROVED** [review]

File: docs/usage.md [hunk 5]
```diff
@@ -75,41 +75,41 @@ First register the populator as a service:
...
-      target: YourNamespace\Person
-      populators:
-        - YourNamespace\PersonNamePopulator
-        # additional populators may follow
+    converter:
+        person.converter:
+            target: YourNamespace\Person
+            populators:
+                - YourNamespace\PersonNamePopulator
+                # additional populators may follow
```
Comment (jdreesen) [inline docs/usage.md]: Please revert these unnecessary indentations in this file.

File: docs/usage.md [hunk 6]
```diff
@@ -220,8 +220,9 @@ Conversion done.
 
 ## Special Populators
 
-After working a while with the converter pattern, you will notice that many scenarios in the population are very similar.
-If the source property can be copied directly to the target property, but only the names of the properties change, 
+After working a while with the converter pattern, you will notice that many scenarios in the population are very
+similar.
```
Comment (jdreesen) [inline docs/usage.md]: Please don't add these strange line breaks.

```suggestion
After working a while with the converter pattern, you will notice that many scenarios in the population are very similar.
```

File: src/Populator/PropertyMappingPopulator.php [hunk 7]
```diff
@@ -43,7 +43,13 @@ public function __construct(
...
     {
         try {
-            $value = $this->accessor->getValue($source, $this->sourceProperty) ?? $this->defaultValue;
+            if ('$this' !== $this->sourceProperty) {
+                $sourcePropertyValue = $this->accessor->getValue($source, $this->sourceProperty);
+            } else {
+                $sourcePropertyValue = $source;
+            }
+
+            $value = $sourcePropertyValue ?? $this->defaultValue;
```
Comment (jdreesen) [inline src/Populator/PropertyMappingPopulator.php]: As `$source` cannot be `null`, I'd prefer writing it like this:

```suggestion
            $value = '$this' !== $this->sourceProperty
            ? $this->accessor->getValue($source, $this->sourceProperty) ?? $this->defaultValue
            : $source;
```

**Review by jdreesen — APPROVED** [review]

**Pull request merged**